zipline/data/data_portal.py

```diff
@@ -59,6 +59,362 @@
 HISTORY_FREQUENCIES = set(["1m", "1d"])
 
 
+class DailyHistoryAggregator(object):
+    """
+    Converts minute pricing data into a daily summary, to be used for the
+    last slot in a call to history with a frequency of `1d`.
+
+    This summary is the same as a daily bar rollup of minute data, with the
+    distinction that the summary is truncated to the `dt` requested.
+    i.e. the aggregation slides forward during a the course of simulation day.
+
+    Provides aggregation for `open`, `high`, `low`, `close`, and `volume`.
+    The aggregation rules for each price type is documented in their respective
+
+    """
+
+    def __init__(self, market_opens, minute_reader):
+        self._market_opens = market_opens
+        self._minute_reader = minute_reader
+
+        # The caches are structured as (date, market_open, entries), where
+        # entries is a dict of asset -> (last_visited_dt, value)
+        #
+        # Whenever an aggregation method determines the current value,
+        # the entry for the respective asset should be overwritten with a new
+        # entry for the current dt.value (int) and aggregation value.
+        #
+        # When the requested dt's date is different from date the cache is
+        # flushed, so that the cache entries do not grow unbounded.
+        #
+        # Example cache:
+        # cache = (date(2016, 3, 17),
+        #          pd.Timestamp('2016-03-17 13:31', tz='UTC'),
+        #          {
+        #              1: (1458221460000000000, np.nan),
+        #              2: (1458221460000000000, 42.0),
+        #         })
+        self._opens_cache = None
+        self._highs_cache = None
+        self._lows_cache = None
+        self._closes_cache = None
+        self._volumes_cache = None
+
+        # The int value is used for deltas to avoid extra computation from
+        # creating new Timestamps.
+        self._one_min = pd.Timedelta('1 min').value
+
+    def opens(self, assets, dt):
+        """
+        The open field's aggregation returns the first value that occurs
+        for the day, if there has been no data on or before the `dt` the open
+        is `nan`.
+
+        Once the first non-nan open is seen, that value remains constannt per
+        asset for the remainder of the day.
+
+        Returns
+        -------
+        np.array with dtype=float64, in order of assets parameter.
+        """
+        date = dt.date()
+        dt_value = dt.value
+        if self._opens_cache is None or self._opens_cache[0] != date:
```

Comment:
these ~5 lines are repeated in each of the methods, should we generalize into a helper method?